include/revng/ADT/GenericGraph.h

```diff
@@ -388,6 +393,277 @@ class BidirectionalNode
+    auto Comparator = [&P](auto &Edge) { return Edge.Neighbor == &P; };
+    return std::find_if(Predecessors.begin(), Predecessors.end(), Comparator);
+  }
+  typename EdgeViewContainer::const_iterator
+  findPredecessor(DerivedType const &P) const {
+    auto Comparator = [&P](auto const &Edge) { return Edge.Neighbor == &P; };
+    return std::find_if(Predecessors.begin(), Predecessors.end(), Comparator);
+  }
+
+public:
+  bool hasSuccessor(DerivedType const &S) const {
+    return findSuccessor(S) != Successors.end();
+  }
+  bool hasPredecessor(DerivedType const &S) const {
+    return findPredecessor(S) != Predecessors.end();
+  }
+
+public:
+  size_t successorCount() const { return Successors.size(); }
+  size_t predecessorCount() const { return Predecessors.size(); }
+
+  bool hasSuccessors() const { return Successors.size() != 0; }
+  bool hasPredecessors() const { return Predecessors.size() != 0; }
+
+public:
+  // Maybe this overload should be `protected`. But it's faster than the
+  // alternative, so I'm hesitant.
+  typename EdgeOwnerContainer::iterator
+  removeSuccessor(typename EdgeOwnerContainer::const_iterator SuccessorIt) {
+    // Maybe we should do some checks as to whether `SuccessorIt` is valid.
+
+    auto *Successor = SuccessorIt->Neighbor;
+    auto PredecessorIt = Successor->findPredecessor(*this);
+    revng_assert(PredecessorIt != Successor->Predecessors.end(),
+                 "Half of an edge is missing, graph layout is broken.");
+    Successor->Predecessors.erase(PredecessorIt);
+    return Successors.erase(SuccessorIt);
+  }
+  auto removeSuccessor(DerivedType const &S) {
+    return removeSuccessor(findSuccessor(S));
+  }
+
+  // Maybe this overload should be `protected`. But it's faster than the
+  // alternative, so I'm hesitant.
+  typename EdgeViewContainer::iterator
+  removePredecessor(typename EdgeViewContainer::const_iterator PredecessorIt) {
+    // Maybe we should do some checks as to whether `PredecessorIt` is valid.
+
+    auto *Predecessor = PredecessorIt->Neighbor;
+    auto SuccessorIt = Predecessor->findSuccessor(*this);
+    revng_assert(SuccessorIt != Predecessor->Successors.end(),
+                 "Half of an edge is missing, graph layout is broken.");
+    Predecessor->Successors.erase(SuccessorIt);
+    return Predecessors.erase(PredecessorIt);
+  }
+  auto removePredecessor(DerivedType const &P) {
+    return removePredecessor(findPredecessor(P));
+  }
+
+public:
+  NewNode &disconnect() {
+    for (auto It = Successors.begin(); It != Successors.end(); ++It)
+      It = removeSuccessor(It);
+    for (auto It = Predecessors.begin(); It != Predecessors.end(); ++It)
+      It = removePredecessor(It);
+    return *this;
+  }
+
+private:
+  EdgeOwnerContainer Successors;
+  EdgeViewContainer Predecessors;
+};
+
+template<typename T>
+concept IsNewNode = requires {
+  T::is_new_node;
+  typename llvm::Inverse<T *>;
```

Comment:
Uhm, why?

Reply:
It's the same concept used for both of the existing node types. The name is weird, I know, but that's just what the node is currently named.

Reply:
Like [here](https://github.com/Cvelth/revng/blob/5d97b2288372fd249c1338ed973d4200ef9f304a/include/revng/ADT/GenericGraph.h#L397), for example.